### PR TITLE
Add Liquid lexer

### LINKED ISSUE #185: Add Liquid Lexer
Hi! I would love to see a lexer for Liquid. One is in pygments.

- `LiquidLexer` class in Pygments: https://bitbucket.org/birkenfeld/pygments-main/src/53092018461a95c3310adc276f1ef7f6d7715b50/pygments/lexers/templates.py#cl-1857
- Changeset: https://bitbucket.org/birkenfeld/pygments-main/commits/53092018461a
- Original issue: https://bitbucket.org/birkenfeld/pygments-main/issue/977/add-liquid-lexer
- Pygments lexer by @maul-esel: https://gist.github.com/maul-esel/6652045

Not sure how to go about converting this to Rouge but wanted to open the issue so we can track it.

### PR DESCRIPTION
**Pull request by RadLikeWhoa**
This PR adds a new lexer for the liquid templating language, complete with demo and visual spec. It would close #185.

The lexer is mostly a 1:1 port of [Pygment's liquid lexer](https://bitbucket.org/birkenfeld/pygments-main/src/53092018461a95c3310adc276f1ef7f6d7715b50/pygments/lexers/templates.py#cl-1857), but I've improved some bits of the code. However, someone should probably look at the code again as I don't know Ruby that well.

### CODE REVIEW

**Comment by jneen** [comment]
Awesome, thanks!  I have one or two stylistic nits, but I'd love to merge this.

File: lib/rouge/demos/liquid [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  {% for product in products %}
+    <li>
+      <h2>{{ product.title }}</h2>
+      Only {{ product.price | format_as_money }}
+
+      <p>{{ product.description | prettyprint | truncate: 200  }}</p>
+
+    </li>
+  {% endfor %}
+</ul>
```
Comment (jneen) [inline lib/rouge/demos/liquid]: Could you make sure all the files have newlines at the end?

File: lib/rouge/lexers/liquid.rb [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+          push :tagBlock
+        end
+        rule /(\{\{)(\s*)/ do
+          groups Punctuation, Text::Whitespace
+          push :output
+        end
+        rule /\{/, Text
+      end
+
+      state :tagBlock do
```
Comment (jneen) [inline lib/rouge/lexers/liquid.rb]: The convention here is to use `snake_case` for symbols, if you could convert these that'd be awesome.

File: lib/rouge/lexers/liquid.rb [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        rule (/(else)(\s*)(%\})/) do
+          groups Keyword::Reserved, Text::Whitespace, Punctuation
+          pop!
+        end
+
+        # other builtin blocks
+        rule (/(capture)(\s+)([^\s%]+)(\s*)(%\})/) do
+          groups Name::Tag, Text::Whitespace, Name::Attribute, Text::Whitespace, Punctuation
+          pop!
+        end
```
Comment (jneen) [inline lib/rouge/lexers/liquid.rb]: newlines between these blocks would also be nice.

File: lib/rouge/lexers/liquid.rb [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        rule (/(raw)(\s*)(%\})/) do
+          groups Name::Tag, Text::Whitespace, Punctuation
+          push :raw
+        end
+        rule (/(assign)(\s*)/) do
+          groups Name::Tag, Text::Whitespace
+          push :assign
+        end
+        rule (/(include)(\s*)/) do
```
Comment (jneen) [inline lib/rouge/lexers/liquid.rb]: Is there a reason you have to add `\s*` to the end here?  It's generally a little nicer just to mixin `:whitespace` to whatever state is next imo.

File: lib/rouge/lexers/liquid.rb [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+      end
+
+      state :tagBlock do
+        # builtin logic blocks
+        rule /(if|unless|elsif|case)(?=\s+)/, Keyword::Reserved, :condition
+        rule /(when)(\s+)/ do
+          groups Keyword::Reserved, Text::Whitespace
+          push :when
+        end
+        rule (/(else)(\s*)(%\})/) do
```
Comment (jneen) [inline lib/rouge/lexers/liquid.rb]: You can omit the parens as long as you're not using a curly-brace block.

**Comment by jneen** [comment]
Other than that, everything looks great :+1:.  I'm also happy to look at some of these issues if you're crunched for time.

**Comment by RadLikeWhoa** [comment]
@jneen Thanks for the comments. I'm pretty sure I fixed all the issues now. :)

**Comment by jneen** [comment]
OK.  I still think there are more `\s*`'s than necessary, but it's ok to merge for now.

**Comment by jneen** [comment]
Released in 1.8.0.

**Pull request merged**